### PR TITLE
[Statistics] Introduce Community Statistics Reports

### PR DESCRIPTION
**Pull request by loic-sharma**
# Introduce Community Statistics Reports

This change introduces the concept of "community" statistics report that will be displayed by default. These reports filter out packages owned by Microsoft and .NET Framework. Users can choose to "Show all packages" to view reports that include Microsoft's and .NET Framework's packages. With this change, packages like `System.Globalization` will no longer clutter our default statistic reports! 😄 

Note: The `JsonStatisticsService` had some serious concurrency issue and did not cache reports that it loaded from blob storage. This pull request fixes that by loading reports once an hour.

This change depends on https://github.com/NuGet/NuGet.Jobs/pull/239.
Fixes https://github.com/NuGet/NuGetGallery/issues/3417.

# Screenshots

## Statistics Homepage

one

## Detailed Report

Clicking on the "Show more..." links at the bottom of the homepage will take you to the following reports:

### Most Downloaded Packages

most-downloaded-packages

### Most Downloaded Package Versions

most-downloaded-package-versions

### CODE REVIEW

**Comment by anangaur** [comment]
I dont think we should have the filter on the https://www.nuget.org/stats page. The filter should only be there on the details pages:
https://www.nuget.org/stats/packageversions
https://www.nuget.org/stats/packages

**Comment by anangaur** [comment]
Something like this:

image

**Comment by loic-sharma** [comment]
Added a filter to both index and detailed pages. I will update the screenshots on Monday.

File: src/NuGetGallery/Services/JsonStatisticsService.cs [hunk 1]
```diff
@@ -109,39 +173,58 @@ public IEnumerable<StatisticsWeeklyUsageItem> Last6Weeks
             }
         }
 
-        public async Task<StatisticsReportResult> LoadDownloadPackageVersions()
+        private Task<StatisticsReportResult> LoadDownloadPackageVersions()
+        {
+            var statisticsPackagesItemViewModels = (List<StatisticsPackagesItemViewModel>)DownloadPackageVersionsAll;
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/JsonStatisticsService.cs]: TODO: Delete this.

File: src/NuGetGallery/Services/StatisticsReportName.cs [hunk 2]
```diff
@@ -18,6 +18,21 @@ public enum StatisticsReportName
...
+        /// <summary>
+        /// Most frequently downloaded community package, specific to actual version.
+        /// This excludes Microsoft and ASP.NET packages.
+        /// </summary>
+        RecentCommunityPopularityDetail,
+        /// <summary>
+        /// Breakout by version for a community package (drill down from RecentPopularity).
+        /// This excludes Microsoft and ASP.NET packages.
+        /// </summary>
+        RecentCommunityPopularityDetail_,
```
Comment (loic-sharma) [inline src/NuGetGallery/Services/StatisticsReportName.cs]: TODO: This is not needed, remove.

File: src/NuGetGallery/Controllers/ApiController.cs [hunk 3]
```diff
@@ -704,9 +704,9 @@ protected internal virtual Stream ReadPackageFromRequest()
         [ActionName("StatisticsDownloadsApi")]
         public virtual async Task<ActionResult> GetStatsDownloads(int? count)
         {
-            var result = await StatisticsService.LoadDownloadPackageVersions();
+            await StatisticsService.Refresh();
```
Comment (joelverhagen) [inline src/NuGetGallery/Controllers/ApiController.cs]: Why this change?

File: src/NuGetGallery/Views/Statistics/Index.cshtml [hunk 4]
```diff
@@ -16,16 +16,21 @@
...
 <section role="main" class="container main-container page-statistics-overview">
     <h1>Statistics</h1>
+
     <span class="page-subheading ms-fontSize-l">
         @Html.Partial("_LastUpdated", Model)
     </span>
-
+    <p>
+        Show all packages:
+        <input type="checkbox" data-bind="checked: showAllPackageDownloads" />
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Statistics/Index.cshtml]: The include prerelease checkbox on [search](https://www.nuget.org/packages) and the [per-package stats page](https://www.nuget.org/stats/packages/Newtonsoft.Json?groupby=Version&groupby=ClientName) both are linkable. Meaning clicking on them updates the URL in the address bar so people can link to the results. You should do this either with a `GET` form or `replaceState` in JavaScript.

**Comment by joelverhagen** [comment]
Do users really care about the old report? Is this driven by a specific customer ask? It is unclear from the linked issue if we should provide a way to toggle between them.

**Comment by anangaur** [comment]
The filter (check box) should be enabled by default to filter out the Microsoft packages as that is what users want. It serves 2 purposes:
1. Soft landing for users who are used to seeing the Microsoft packages in the list.
2. Messaging: We need to tell what is it that we show on the page :)

File: src/NuGetGallery/Views/Statistics/Index.cshtml (hunk 4 above)
Comment (loic-sharma) [reply src/NuGetGallery/Views/Statistics/Index.cshtml]: Good suggestion. However, due to feedback this will no longer be linkable.

File: src/NuGetGallery/Controllers/ApiController.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGetGallery/Controllers/ApiController.cs]: See the description:

> The JsonStatisticsService had some serious concurrency issue and did not cache reports that it loaded from blob storage. This pull request fixes that by loading reports once every four hours.

**Comment by loic-sharma** [comment]
Addressed feedback.

File: src/NuGetGallery/Controllers/ApiController.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGetGallery/Controllers/ApiController.cs]: To elaborate a little: all the `Load...` reports were merged into the `Refresh` method.

File: src/NuGetGallery/Controllers/StatisticsController.cs [hunk 5]
```diff
@@ -73,27 +75,23 @@ public StatisticsController(IStatisticsService statisticsService, IAggregateStat
...
+            await _statisticsService.Refresh();
 
             var model = new StatisticsPackagesViewModel
             {
-                IsDownloadPackageAvailable = availablity[0].Loaded,
-                DownloadPackagesSummary = _statisticsService.DownloadPackagesSummary,
-                IsDownloadPackageDetailAvailable = availablity[1].Loaded,
-                DownloadPackageVersionsSummary = _statisticsService.DownloadPackageVersionsSummary,
-                IsNuGetClientVersionAvailable = availablity[2].Loaded,
+                IsDownloadPackageAvailable = _statisticsService.DownloadCommunityPackagesResult.Loaded,
```
[the remark this answers is not in the document]
Comment (xavierdecoster) [reply src/NuGetGallery/Controllers/StatisticsController.cs]: Same for the `Loaded` property which currently sounds like an event: `IsLoaded`?

[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/NuGetGallery/Controllers/StatisticsController.cs]: Will do! :)
